### PR TITLE
Added optional num parameter for dataset generation

### PR DESCRIPTION
**Pull request by sahil-springworks**
### Summary

Added a new optional `num` parameter to the dataset `generate_questions_from_nodes` function allowing users to fetch only "num" number of questions from the given nodes.

### CODE REVIEW

**Review by Disiok — CHANGES_REQUESTED** [review]
Thanks for the contribution! Minor comment on cleaner implementation of this

File: llama_index/evaluation/dataset_generation.py [hunk 1]
```diff
@@ -112,11 +112,15 @@ def from_documents(
...
 
-    def _node_question_generator(self, nodes: List[Node]) -> List[str]:
+    def _node_question_generator(
+        self, nodes: List[Node], num: Optional[int] = None
+    ) -> List[str]:
         """Node question generator."""
         questions = []
 
         for node in nodes:
+            if num != None and len(questions) >= num:
```
[the remark this answers is not in the document]
Comment (sahil-springworks) [reply llama_index/evaluation/dataset_generation.py:122]: Actually `num` is not the number of nodes that need to be evaluated but instead the number of questions that need to be generated so we need to check for the length of the questions.

[the remark this answers is not in the document]
Comment (Disiok) [reply llama_index/evaluation/dataset_generation.py:122]: Oh I see, each iteration can generate multiple questions. Sorry totally missed this!

Comment (Disiok) [inline llama_index/evaluation/dataset_generation.py:122]: very nit: more idiomatic in Python to do `num is not None`

**Review by Disiok — APPROVED** [review]

**Comment by Disiok** [comment]
Thanks! Happy to merge after fixing the nit and passing CI

**Comment by sahil-springworks** [comment]
> Thanks! Happy to merge after fixing the nit and passing CI

Made the change to use `num is not None`

File: llama_index/evaluation/dataset_generation.py [hunk 2]
```diff
@@ -136,8 +140,10 @@ def _node_question_generator(self, nodes: List[Node]) -> List[str]:
 
         questions = [question for question in questions if question != ""]
 
+        if num != None:
```
Comment (jerryjliu) [inline llama_index/evaluation/dataset_generation.py:143]: can you do `num is not None` here too

**Review by jerryjliu — APPROVED** [review]
approving with one nit

File: llama_index/evaluation/dataset_generation.py (hunk 2 above)
Comment (sahil-springworks) [reply llama_index/evaluation/dataset_generation.py]: > can you do `num is not None` here too

Sure, sorry to have missed that. Done.

**Review by jerryjliu — APPROVED** [review]

**Comment by sahil-springworks** [comment]
@Disiok @jerryjliu any action needed from my side for the merge of this PR?

**Comment by logan-markewich** [comment]
@sahil-springworks looks like mypy is noticing a missing type annotation here. With that fixed it should be good to go

`llama_index/evaluation/dataset_generation.py:119: error: Need type annotation for "questions" (hint: "questions: List[<type>] = ...")  [var-annotated]`

**Comment by sahil-springworks** [comment]
@logan-markewich added. Thanks

**Comment by logan-markewich** [comment]
Thanks!!

**Pull request merged**